mds: fix deadlock between unlinking and linkmerge

lxbsz opened the pull request:
In case of one inode have multiple hardlinks, and if an unlink
request removes the primary dentry and just before the linkmerge
is triggered a new unlink request comes. The new unlink request
will need the linkmerge to finish.

We should just wait the linkmerge to finish before handling the
new unlink request for the same inode.

Fixes: https://tracker.ceph.com/issues/58340
Fixes: https://tracker.ceph.com/issues/61818





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

dparmar18 commented on src/mds/Server.cc:
```diff
@@ -2556,6 +2556,12 @@ void Server::handle_client_reply(const cref_t<MClientReply> &reply)
 
+      // move the state back to STATE_UNLINKING
+      if (dn->state_test(CDentry::STATE_UNLINK_REINTEGRATING)) {
+        dn->state_set(CDentry::STATE_UNLINKING);
+        dn->state_clear(CDentry::STATE_UNLINK_REINTEGRATING);
```
shouldn't we first clear the state and then set?

lxbsz replied:
> shouldn't we first clear the state and then set?

This should be okay since the it always holding the global `mds_lock`.
But I will fix this anyway. 

Thanks @dparmar18

batrick reviewed (changes requested):
I think there is a simpler fix here. The core problem is that `StrayManager::reintegrate_stray` sets the dentry state to `REINTEGRATING` before the rename op is even dispatched and acquired the necessary locks. Until that happens, we've simply recorded an intent (the op) without actually changing anything **or acquiring any locks**.

Had `Server::handle_client_rename` set the reintegration state after acquiring locks and checking for any pending unlinks, we would not have hit this deadlock, no?

lxbsz commented:
> I think there is a simpler fix here. The core problem is that `StrayManager::reintegrate_stray` sets the dentry state to `REINTEGRATING` before the rename op is even dispatched and acquired the necessary locks. Until that happens, we've simply recorded an intent (the op) without actually changing anything **or acquiring any locks**.
> 
> Had `Server::handle_client_rename` set the reintegration state after acquiring locks and checking for any pending unlinks, we would not have hit this deadlock, no?

Not sure I totally get it.

The root cause is that the `reintegrate` is triggered by the `unlink` request itself. I found a simper approach, that is skipping checking the `is_unlink_pending()` or always return true from `is_unlink_pending()` if the `rename` is a linkmerge request.

lxbsz commented:
@batrick I have simplify the code, please take a look again. Thanks.

lxbsz commented:
jenkins test api

batrick commented:
Also, I'm talking about this code:

https://github.com/ceph/ceph/blob/a4b76efd31bc0036d7a88cca0184e45f0d58d0f0/src/mds/StrayManager.cc#L700

lxbsz commented:
> Also, I'm talking about this code:
> 
> https://github.com/ceph/ceph/blob/a4b76efd31bc0036d7a88cca0184e45f0d58d0f0/src/mds/StrayManager.cc#L700

@batrick I have reverted the previous commit, which introduced the `STATE_REINTEGRATING`. And fixed these two issues in a following commit https://github.com/ceph/ceph/pull/52199/commits/9d8f88d06cfaf6e64ee5dce980c7a122d4dacc67, which will always wait the previous unlink's linkmerge to finish.

Please review again. Thanks
- Xiubo

batrick commented on src/mds/Server.cc:
```diff
@@ -7929,6 +7929,23 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
+
+  // wait the previous unlink finshes thoroughly.
+  CDentry* _dn = in->get_projected_parent_dn();
+  if (_dn->get_dir()->inode->is_stray()) {
```
So I have a little time to comment on this tonight. Do you expect that `rdlock_path_xlock_dentry` will trigger a stray reintegration (I cannot recall off-hand and the code is spaghetti).  I would be clear that we're not waiting for "unlink finishes thoroughly" but that even acquiring the necessary locks on the dentry will automatically cause the rename to be fired off, which we should wait for.

If that rename is **not** fired off by `rdlock_path_xlock_dentry`, I would be concerned this unlink could wait forever if nothing else touches that dentry to trigger reintegration.

batrick commented on src/mds/Server.cc:
```diff
@@ -7929,6 +7929,23 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
+    // after reintegrating retry it
+    in->add_waiter(CInode::WAIT_UNLINK, new C_MDS_RetryRequest(mdcache, mdr));
+
+    // trigger to flush the logs, after which will it trigger stray dentry reintegrate
```
how exactly does flushing the logs trigger?

batrick commented on src/mds/Server.cc:
```diff
@@ -7929,6 +7929,23 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
+  CInode *in = dnl->get_inode();
+
+  // wait the previous unlink finshes thoroughly.
+  CDentry* _dn = in->get_projected_parent_dn();
```
Keep in mind that the parent dn may not be projected at all.

batrick replied:
> This will try to get the primary dentry only, which is possibly still projected after the previous unlink request.

Understood, but my point is that if the CInode is not projected, then I do not believe the journal can be relied on to trigger reintegration?

>From my reading from the code only when the primary dentry is unlinked should it trigger reintegrate ?

I'm not sure to be honest. That's one reason I'm concerned about waiting on a queue.

lxbsz replied:
> > This will try to get the primary dentry only, which is possibly still projected after the previous unlink request.
> 
> Understood, but my point is that if the CInode is not projected, then I do not believe the journal can be relied on to trigger reintegration?

As you know the `get_projected_parent_dn()` will always return a `primary` dentry even not projected.

> 
> > From my reading from the code only when the primary dentry is unlinked should it trigger reintegrate ?
> 
> I'm not sure to be honest. That's one reason I'm concerned about waiting on a queue.

If my understanding is correct, your concern should be the previous `unlink's` projected parent `stray` dentry has possibly been popped from projected list and later a new `link` request comes and projects it, or there have multiple `projected parent` dentries in the projected list and the newest one is not the one we want ?

This won't happen because the `link` request will wait for the `linkmerge` to finish, please see https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L7385-L7396.

And for the `rename` request in case the `srcdn` is a remote dn for the same `CInode` it won't touch the `parent` or the `projected parent` for the `CInode`. While for the case the `dstdn` also the same. Because before the `linkmerge` to finish the `parent` or `projected parent` must be a `stray` dentry.

Did I miss something here ?

lxbsz commented:
jenkins test make check

vshankar commented:
Tests run in ~2 hours - https://shaman.ceph.com/builds/ceph/wip-vshankar-testing-20230828.044427/

(if teuthology behaves :)

batrick commented on src/mds/Server.cc:
```diff
@@ -8008,8 +7929,20 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
+  CInode *in = dnl->get_inode();
+
+  // wait the previous unlink finshes thoroughly.
+  CDentry* _dn = in->get_projected_parent_dn();
+  if (_dn->get_dir()->inode->is_stray()) {
```
```suggestion
    auto* primary_diri = in->get_projected_parent_dir()->inode;
    if (primary_diri->is_stray()) {
```

lxbsz replied:
Will fix it.

batrick commented on src/mds/Server.cc:
```diff
@@ -8008,8 +7929,20 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
+    in->add_waiter(CInode::WAIT_UNLINK, new C_MDS_RetryRequest(mdcache, mdr));
+
+    // trigger to flush the logs, after which will it trigger stray dentry reintegrate
+    mdlog->flush();
```
Okay, hypothetical:

- MDS has nothing in cache
- client opens hard linked file, the primary dentry
- it unlinks that dentry, the remote dentry linking to it is not in cache so no stray migration is triggered
- optional: wait for journal flush so inode is no longer projected
- client tries to unlink the remote dentry (found via `rdlock_path_xlock_dentry`)

Will stray reintegration happen so that the unlink will eventually complete? (I ask because you believe that `rdlock_path_xlock_dentry` will not trigger stray reintegration.)

This type of situation for the code is not uncommon (it's one reason we have request deadlock in the MDS). I'm always leery of seeing a request wait like this without being certain some outstanding operation (which we can clearly point to) will wake it up.

batrick replied:
To be clear: what I'm hoping to see is some code path we can be certain will be hit when the `CInode` is no longer projected. How can we be sure stray reintegration will be triggered (if not already)?

lxbsz replied:
The `link` request is doing the same with here, please see https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L7385-L7396.

The `mdlog` flushing just to make sure that the previous `unlink` request's journal will be flushed and then the `_unlink_local() --> _unlink_local_finish() --> mdcache->notify_stray(straydn) --> ... --> StrayManager::_eval_stray() --> StrayManager::_eval_stray_remote() --> reintegrate_stray()` will be called.

And in `StrayManager::_eval_stray_remote()` it will pick a `remote` dentry arbitrarily:

```
640   /* If no remote_dn hinted, pick one arbitrarily */
641   if (remote_dn == NULL) {
642     if (!stray_in->remote_parents.empty()) {                                                                                                                                                                                       
643       for (const auto &dn : stray_in->remote_parents) {
644         if (dn->last == CEPH_NOSNAP && !dn->is_projected()) {
645           if (dn->is_auth()) { 
646             remote_dn = dn;    
647             if (remote_dn->dir->can_auth_pin()) 
648               break;
649           } else if (!remote_dn) {
650             remote_dn = dn;    
651           }
652         }
653       }
654     }
```

The reintegration will load the remote dentry when needed. And after the reintegration finishing it will wake up the waiters.

I know maybe the reintegration could be triggered by other cases, but that won't be our concern here. Isn't it ?

batrick commented on src/mds/Server.cc:
```diff
@@ -7928,6 +7928,23 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
+
+  // wait the previous unlink finshes thoroughly.
+  auto* primary_diri = in->get_projected_parent_dir()->inode;
+  if (primary_diri->is_stray()) {
```
Okay I've done some extensive testing today and have been unable to ever get this code path to be taken. The main issue is that the client will always `lookup` the file path before unlink. When that unlink request is cleaned up and unlocked, the Locker will `eval_remote` to initiate any reintegration. That will likely occur before the followup unlink request can be issued. Consequently, the rename acquires the locks that the `unlink` RPC needs so it's blocked until that rename completes. By that time, the `primary_diri` is no longer a stray.

I'm not even sure what the original bug is from https://tracker.ceph.com/issues/58340 because that log has long been garbage collected. Of course, I'm fairly confident you figured it out Xiubo. Do you have a reliable test case to share?

Also, I've been looking at #47399 which introduced `WAIT_UNLINK`. After having looked at all of these PRs in depth, that original issue

https://tracker.ceph.com/issues/56695

indicates to me that the problem is 

https://github.com/ceph/ceph/blob/552b6d852cbc7384a05805989a675de7d6b4a3fd/src/mds/Server.cc#L7385

in that this is the **wrong** parent to look at. The parent dn of the inode is irrelevant. What matters is that the target dn's parent is in the same subvolume. This is precisely because the actual inode may temporarily be located in the stray directory pending reintegration. So I *think* that the right fix for issue 56695 is to simply check the snap realm of `srcdn` (which would be gotten from `rdlock_two_paths_xlock_destdn`). Alternatively, don't return EXDEV if the inode's parent dir is a stray.

What do you think? I'm leery of WAIT_UNLINK for the same reasons as WAIT_REINTEGRATE.

lxbsz replied:
> Okay I've done some extensive testing today and have been unable to ever get this code path to be taken. The main issue is that the client will always `lookup` the file path before unlink. When that unlink request is cleaned up and unlocked, the Locker will `eval_remote` to initiate any reintegration. That will likely occur before the followup unlink request can be issued. Consequently, the rename acquires the locks that the `unlink` RPC needs so it's blocked until that rename completes. By that time, the `primary_diri` is no longer a stray.
> 

From my reading the previous test failure logs, it seems the first `unlink` hasn't triggered the `linkmerge` when MDS received a second `unlink`. The test cases it will be something like:

Primary dentry path : /s1/d1/c1/file
Remote dentry path:  /s2/d2/c2/file_link1
Remote dentry path:  /s3/d3/c3/file_link2

Client issued the `unlink` for the primary denty, and then a second `unlink` for one of the remote dentry.

I didn't see why the lookup need to wait the first `unlink` request must release the non-read locks. Just after the early reply the `unlink` request will release the `rdlocks`, won't that enough ?

A detail example please see https://tracker.ceph.com/issues/61818#note-2.

> I'm not even sure what the original bug is from https://tracker.ceph.com/issues/58340 because that log has long been garbage collected. Of course, I'm fairly confident you figured it out Xiubo. Do you have a reliable test case to share?
> 

Before I reproduce this with the `fsstress` test script. And also could see the test failure in `pjd` test script. I think you can just run the `fsstress` script to have a try.

> Also, I've been looking at #47399 which introduced `WAIT_UNLINK`. After having looked at all of these PRs in depth, that original issue
> 
> https://tracker.ceph.com/issues/56695
> 
> indicates to me that the problem is
> 
> https://github.com/ceph/ceph/blob/552b6d852cbc7384a05805989a675de7d6b4a3fd/src/mds/Server.cc#L7385
> 
> in that this is the **wrong** parent to look at. The parent dn of the inode is irrelevant. What matters is that the target dn's parent is in the same subvolume. This is precisely because the actual inode may temporarily be located in the stray directory pending reintegration. So I _think_ that the right fix for issue 56695 is to simply check the snap realm of `srcdn` (which would be gotten from `rdlock_two_paths_xlock_destdn`). Alternatively, don't return EXDEV if the inode's parent dir is a stray.
> 
> What do you think? I'm leery of WAIT_UNLINK for the same reasons as WAIT_REINTEGRATE.

Checked the original issue and the mds code again, yeah, this approach seems won't works because even the `linkmerge` is not triggered yet the `link` still could success and won't depend on it.

Let me update this PR to revert all the previous patches and run the test again.

Thanks @batrick

vshankar commented:
Currently running tests: https://pulpito.ceph.com/vshankar-2023-08-31_04:13:51-fs-wip-vshankar-testing-20230830.153114-testing-default-smithi/

gregsfortytwo commented:
Are we okay to simply remove the WAIT_UNLINK etc flags, now that those are in running clusters?

lxbsz commented:
> Are we okay to simply remove the WAIT_UNLINK etc flags, now that those are in running clusters?

Is that possible that a cluster will run different MDS versions ? If no IMO it's okay since it's only a internal flag in MDS.

vshankar commented:
> Currently running tests: https://pulpito.ceph.com/vshankar-2023-08-31_04:13:51-fs-wip-vshankar-testing-20230830.153114-testing-default-smithi/

This is a pretty rough run - infra issues mostly. I would like those to get resolved first and get a run next.

vshankar commented:
> > Currently running tests: https://pulpito.ceph.com/vshankar-2023-08-31_04:13:51-fs-wip-vshankar-testing-20230830.153114-testing-default-smithi/
> 
> This is a pretty rough run - infra issues mostly. I would like those to get resolved first and get a run next.

I believe the infra failures (failed ssh connection) hasn't been fixed yet, thereby testing for this change is stalled.

gregsfortytwo commented:
> > Are we okay to simply remove the WAIT_UNLINK etc flags, now that those are in running clusters?
> 
> Is that possible that a cluster will run different MDS versions ? If no IMO it's okay since it's only a internal flag in MDS.

We don't make users shut down daemons on minor version upgrades, I think? But my bigger concern was about what happens with these lock states getting encoded in the journal, and what happens when the new daemon replays them. This may just not matter at all — I don't remember how the lock state value is updated/maintained, if the old value will get wiped out when appropriate, or we have a permanently-set bit that nothing reads right now but will cause trouble in the future?

gregsfortytwo commented:
@batrick @vshankar ^

batrick commented:
> > > Are we okay to simply remove the WAIT_UNLINK etc flags, now that those are in running clusters?
> > 
> > 
> > Is that possible that a cluster will run different MDS versions ? If no IMO it's okay since it's only a internal flag in MDS.
> 
> We don't make users shut down daemons on minor version upgrades, I think? But my bigger concern was about what happens with these lock states getting encoded in the journal, and what happens when the new daemon replays them.

These are transient states on the inode. Nothing about this is persisted to RADOS.

Our official MDS upgrade doc is here: https://docs.ceph.com/en/latest/cephfs/upgrading/

It's very clear that you should not upgrade a cluster with max_mds>1. The one exception is the new `fail_fs` flag for cephadm to "fail" the file system so no MDS is active and then upgrade them all at once.

vshankar commented:
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230830.153114

batrick commented:
Going to run this through testing myself too.

vshankar commented:
> Going to run this through testing myself too.

I ran it last week and the (re)run are in comment https://github.com/ceph/ceph/pull/52199#issuecomment-1709512412. Failures are known issues, so this is good for merge.

vshankar commented:
@batrick - https://tracker.ceph.com/projects/cephfs/wiki/Main#11-Sep-2023

lxbsz commented:
> > Going to run this through testing myself too.
> 
> I ran it last week and the (re)run are in comment [#52199 (comment)](https://github.com/ceph/ceph/pull/52199#issuecomment-1709512412). Failures are known issues, so this is good for merge.

@vshankar @batrick Could we merge it now ? Thanks.

batrick commented:
> > > Going to run this through testing myself too.
> > 
> > 
> > I ran it last week and the (re)run are in comment [#52199 (comment)](https://github.com/ceph/ceph/pull/52199#issuecomment-1709512412). Failures are known issues, so this is good for merge.
> 
> @vshankar @batrick Could we merge it now ? Thanks.

I hope to go through the results tomorrow. I've been busy with a customer case today.

batrick commented:
> > > > Going to run this through testing myself too.
> > > 
> > > 
> > > I ran it last week and the (re)run are in comment [#52199 (comment)](https://github.com/ceph/ceph/pull/52199#issuecomment-1709512412). Failures are known issues, so this is good for merge.
> > 
> > 
> > @vshankar @batrick Could we merge it now ? Thanks.
> 
> I hope to go through the results tomorrow. I've been busy with a customer case today.

Here tehy are if you want to have an advance look: https://pulpito.ceph.com/pdonnell-2023-09-09_19:00:20-fs-wip-batrick-testing-20230909.145638-distro-default-smithi/

vshankar commented:
> > > > > Going to run this through testing myself too.
> > > > 
> > > > 
> > > > I ran it last week and the (re)run are in comment [#52199 (comment)](https://github.com/ceph/ceph/pull/52199#issuecomment-1709512412). Failures are known issues, so this is good for merge.
> > > 
> > > 
> > > @vshankar @batrick Could we merge it now ? Thanks.
> > 
> > 
> > I hope to go through the results tomorrow. I've been busy with a customer case today.
> 
> Here tehy are if you want to have an advance look: https://pulpito.ceph.com/pdonnell-2023-09-09_19:00:20-fs-wip-batrick-testing-20230909.145638-distro-default-smithi/

I will have a look today anyway (my test runs show this change is mergeable).

batrick commented:
```
2023-09-09T21:22:58.177+0000 7fc8f05bb700  7 mon.a@0(leader).log v588 update_from_paxos applying incremental log 588 2023-09-09T21:22:57.170653+0000 mds.h (mds.1) 1 : cluster [WRN] 2 slow requests, 2 included below; oldest blocked for > 182.965753 secs
2023-09-09T21:22:58.177+0000 7fc8f05bb700  7 mon.a@0(leader).log v588 journald: cluster
2023-09-09T21:22:58.177+0000 7fc8f05bb700  7 mon.a@0(leader).log v588 update_from_paxos applying incremental log 588 2023-09-09T21:22:57.170658+0000 mds.h (mds.1) 2 : cluster [WRN] slow request 182.965752 seconds old, received at 2023-09-09T21:19:54.204855+0000: client_request(mds.1:81 rename #0x20000000441/c14f #0x61a/2000000087b caller_uid=0, caller_gid=0{}) currently acquired locks
2023-09-09T21:22:58.177+0000 7fc8f05bb700  7 mon.a@0(leader).log v588 journald: cluster
2023-09-09T21:22:58.177+0000 7fc8f05bb700  7 mon.a@0(leader).log v588 update_from_paxos applying incremental log 588 2023-09-09T21:22:57.170661+0000 mds.h (mds.1) 3 : cluster [WRN] slow request 182.965474 seconds old, received at 2023-09-09T21:19:54.205133+0000: client_request(client.15657:25500 unlink #0x20000000441/c14f 2023-09-09T21:19:54.204958+0000 caller_uid=1000, caller_gid=1265{6,36,1000,1265,}) currently requesting remote authpins
```

From: /teuthology/pdonnell-2023-09-09_19:00:20-fs-wip-batrick-testing-20230909.145638-distro-default-smithi/7392394/remote/smithi104/log/df7acbaa-4f54-11ee-9ab7-7b867c8bd7da/ceph-mon.a.log.gz

I'll look more tomorrow but just a heads up @lxbsz

lxbsz commented:
> ```
> 2023-09-09T21:22:58.177+0000 7fc8f05bb700  7 mon.a@0(leader).log v588 update_from_paxos applying incremental log 588 2023-09-09T21:22:57.170653+0000 mds.h (mds.1) 1 : cluster [WRN] 2 slow requests, 2 included below; oldest blocked for > 182.965753 secs
> 2023-09-09T21:22:58.177+0000 7fc8f05bb700  7 mon.a@0(leader).log v588 journald: cluster
> 2023-09-09T21:22:58.177+0000 7fc8f05bb700  7 mon.a@0(leader).log v588 update_from_paxos applying incremental log 588 2023-09-09T21:22:57.170658+0000 mds.h (mds.1) 2 : cluster [WRN] slow request 182.965752 seconds old, received at 2023-09-09T21:19:54.204855+0000: client_request(mds.1:81 rename #0x20000000441/c14f #0x61a/2000000087b caller_uid=0, caller_gid=0{}) currently acquired locks
> 2023-09-09T21:22:58.177+0000 7fc8f05bb700  7 mon.a@0(leader).log v588 journald: cluster
> 2023-09-09T21:22:58.177+0000 7fc8f05bb700  7 mon.a@0(leader).log v588 update_from_paxos applying incremental log 588 2023-09-09T21:22:57.170661+0000 mds.h (mds.1) 3 : cluster [WRN] slow request 182.965474 seconds old, received at 2023-09-09T21:19:54.205133+0000: client_request(client.15657:25500 unlink #0x20000000441/c14f 2023-09-09T21:19:54.204958+0000 caller_uid=1000, caller_gid=1265{6,36,1000,1265,}) currently requesting remote authpins
> ```
> 
> From: /teuthology/pdonnell-2023-09-09_19:00:20-fs-wip-batrick-testing-20230909.145638-distro-default-smithi/7392394/remote/smithi104/log/df7acbaa-4f54-11ee-9ab7-7b867c8bd7da/ceph-mon.a.log.gz
> 
> I'll look more tomorrow but just a heads up @lxbsz

This was caused by https://github.com/ceph/ceph/pull/46331 and I also reverted the corresponding commits.

vshankar commented:
> [Edit] This was caused by #46331 and I also reverted the corresponding commits. Since the `async dirop` is disabled in downstream, and just reverting these commits will be safe, I will fix the old issue again after this.

I misunderstood what you said @lxbsz when you pinged me separately regarding this. I get it now (sorry - too many PRs to look at).

batrick commented:
https://tracker.ceph.com/projects/cephfs/wiki/Main#2023-Sep-12

batrick commented:
@lxbsz the changes associated with this PR are a bit tangled. Can you please go through all the backports for the relevant reverts and close or update them (that they will soon be reverted when this is backported). Make sure the tracker tickets are coherent and consistent.

```
batrick@menzoberranzan ~/ceph$ src/script/ptl-tool.py --base main --merge-branch-name main --branch HEAD 52199
Will merge PRs: [52199]
Detaching HEAD onto base: main
Merging PR #52199
[ 5b9df823 ] Ceph tracker cited: Fixes: https://tracker.ceph.com/issues/56695
[ 8e240303 ] Ceph tracker cited: Fixes: https://tracker.ceph.com/issues/61818
[ 4ec2b03f ] Ceph tracker cited: Fixes: https://tracker.ceph.com/issues/61818
[ b0b279ae ] Ceph tracker cited: Fixes: https://tracker.ceph.com/issues/61818
[ 84c7da72 ] Ceph tracker cited: Fixes: https://tracker.ceph.com/issues/61818
[ a9dd5df2 ] Ceph tracker cited: Fixes: https://tracker.ceph.com/issues/61818
[ https://github.com/ceph/ceph/pull/52199 ] Ceph tracker cited: Fixes: https://tracker.ceph.com/issues/58340
[ https://github.com/ceph/ceph/pull/52199 ] Ceph tracker cited: Fixes: https://tracker.ceph.com/issues/61818
[ https://github.com/ceph/ceph/pull/52199 ] Ceph tracker cited:   - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
[ https://github.com/ceph/ceph/pull/52199 ] Ceph tracker cited:   - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
[ https://github.com/ceph/ceph/pull/52199#issuecomment-1713762506 ] Ceph tracker cited: @batrick - https://tracker.ceph.com/projects/cephfs/wiki/Main#11-Sep-2023
[ https://github.com/ceph/ceph/pull/52199#issuecomment-1721527079 ] Ceph tracker cited: https://tracker.ceph.com/projects/cephfs/wiki/Main#2023-Sep-12
[ https://github.com/ceph/ceph/pull/52199#discussion_r1309203317 ] Ceph tracker cited: I'm not even sure what the original bug is from https://tracker.ceph.com/issues/58340 because that log has long been garbage collected. Of course, I'm fairly confident you figured it out Xiubo. Do you have a reliable test case to share?
[ https://github.com/ceph/ceph/pull/52199#discussion_r1309203317 ] Ceph tracker cited: https://tracker.ceph.com/issues/56695
[ https://github.com/ceph/ceph/pull/52199#discussion_r1309652906 ] Ceph tracker cited: A detail example please see https://tracker.ceph.com/issues/61818#note-2.
[ https://github.com/ceph/ceph/pull/52199#discussion_r1309652906 ] Ceph tracker cited: > I'm not even sure what the original bug is from https://tracker.ceph.com/issues/58340 because that log has long been garbage collected. Of course, I'm fairly confident you figured it out Xiubo. Do you have a reliable test case to share?
[ https://github.com/ceph/ceph/pull/52199#discussion_r1309652906 ] Ceph tracker cited: > https://tracker.ceph.com/issues/56695
Leaving HEAD detached; no branch anchors your commits
```

lxbsz commented:
> @lxbsz the changes associated with this PR are a bit tangled. Can you please go through all the backports for the relevant reverts and close or update them (that they will soon be reverted when this is backported). Make sure the tracker tickets are coherent and consistent.
> 

@batrick Sure, I will do that today. Thanks!
